SDK banner tutorials (#9696)
braze-inc/braze-docs: _docs/_developer_guide/banners/tutorials/integrating_banners.md

        return true
    }
}
```

```swift file=SampleApp.swift
import SwiftUI

@main
struct SampleApp: App {
    // Bind the AppDelegate into the SwiftUI lifecycle
    @UIApplicationDelegateAdaptor(AppDelegate.self) var delegate

    var body: some Scene {
        WindowGroup {
            BannerSwiftUIView()
        }
    }
}
```

```swift file=BannerSwiftUIView.swift
import BrazeKit
import BrazeUI
import SwiftUI

@available(iOS 13.0, *)
struct BannerSwiftUIView: View {

  static let bannerPlacementID = "top-1"

  @State var hasBannerForPlacement: Bool = false
  @State var contentHeight: CGFloat = 0

  var body: some View {
    VStack {
      Text("Your Content Here")
        .frame(maxWidth: .infinity, maxHeight: .infinity)
      if let braze = AppDelegate.braze,
        hasBannerForPlacement
      {
        BrazeBannerUI.BannerView(
          placementId: BannerSwiftUIView.bannerPlacementID,
          braze: braze,
          processContentUpdates: { result in
            switch result {
            case .success(let updates):
              if let height = updates.height {
                self.contentHeight = height
              }

>             case .failure(_):

Reviewer: ```suggestion
            case .failure:
```
This isn't necessary. Maybe thought it could be good to include so readers know they can capture the failure reason if they want?

            android:layout_height="wrap_content"
            app:placementId="top-1" />

        <!-- ...the rest of your activity layout -->

    </LinearLayout>
</ScrollView>
```

!!step
lines-MainApplication.kt=12

#### 1. Enable debugging (optional)

Enable debugging while developing to make troubleshooting easier!

!!step
lines-MainApplication.kt=21-28

#### 2. Subscribe to banner updates

Use `subscribeToBannersUpdates()` to listen for and react to incoming banner refresh results from the server.

!!step
lines-MainApplication.kt=30-34

#### 3. Refresh placements

Invoke `requestBannersRefresh()` to fetch the latest banner content from Braze for the specified placement ID(s).

!!step
lines-banners.xml=15-19

#### 4. Define the BannerView in banners.xml

Specify a `<com.braze.ui.banners.BannerView>` element with `app:placementId="<placement-id>"` so that Braze knows where to render the banner content in your UI.

{% endscrolly %}
{% endsdktab %}
{% sdktab swift %}
{% tabs %}
{% tab UIKit %}
{% scrolly %}

```swift file=AppDelegate.swift
import UIKit
import BrazeKit
import BrazeUI

class AppDelegate: UIResponder, UIApplicationDelegate {

>     static var braze: Braze!
> 
>     func application(
>       _ application: UIApplication,
>       didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
>     ) -> Bool {
> 
>         // Braze configuration with your SDK API key and endpoint
>         let configuration = Braze.Configuration(apiKey: "YOUR-API-TOKEN", endpoint: "YOUR-ENDPOINT")
>         configuration.logger.level = .debug
> 
>         // Initialize Braze SDK instance
>         AppDelegate.braze = Braze(configuration: configuration)
> 
>         // Request a banners refresh
>         AppDelegate.braze.banners.requestBannersRefresh(placementIds: ["top-1"])

Reviewer 2: nit: I noticed we treat `braze` as nullable in the other files like `BrazeSwiftUIView.swift`, so this would make it actually be nullable here
```suggestion
    static var braze: Braze? = nil

    func application(
      _ application: UIApplication,
      didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
    ) -> Bool {
        // Braze configuration with your SDK API key and endpoint
        let configuration = Braze.Configuration(apiKey: "YOUR-API-TOKEN", endpoint: "YOUR-ENDPOINT")
        configuration.logger.level = .debug

        // Initialize Braze SDK instance
        AppDelegate.braze = Braze(configuration: configuration)

        // Request a banners refresh
        AppDelegate.braze?.banners.requestBannersRefresh(placementIds: ["top-1"])
```

        return true
    }
}
```

```swift file=SampleApp.swift
import SwiftUI

@main
struct SampleApp: App {
    // Bind the AppDelegate into the SwiftUI lifecycle
    @UIApplicationDelegateAdaptor(AppDelegate.self) var delegate

    var body: some Scene {
        WindowGroup {
            ContentView()
        }
    }
}
```

```swift file=BannerViewController.swift
import UIKit
import BrazeKit
import BrazeUI

final class BannerViewController: UIViewController {

  static let bannerPlacementID = "top-1"

>   var bannerHeightConstraint: NSLayoutConstraint?

Reviewer 2: nit: (Here and the places it is called)
```suggestion
  var bannerHeightConstraints: NSLayoutConstraint?
```

Reviewer 3: We can probably keep this as singular. The height would just be a single value on the view, so this should be okay in its current state IMO!

      bannerHeightConstraint!,
    ])
  }
}
```

!!step
lines-AppDelegate.swift=15

#### 1. Enable debugging (optional)

Enable debugging while developing to make troubleshooting easier!

!!step
lines-AppDelegate.swift=21

#### 2. Trigger a banners refresh

Call `requestBannersRefresh` for your placement ID after initializing Braze.
This proactively fetches the latest banner data, ensuring your app can display new content quickly after launch.

!!step
lines-BannerViewController.swift=18-32

#### 3. Initialize the Braze banner view with a processContentUpdates callback

Create your `BrazeBannerUI.BannerUIView` with the correct placement ID and Braze instance.
The `processContentUpdates` callback will fire whenever the banner height changes.
Inside the callback, unhide the banner and update its height constraint using the provided content height (capped at 80pt for this example).

!!step
lines-BannerViewController.swift=34-36

#### 4. Hide banner by default, enable Auto Layout

Hide the banner view until content loads and disable autoresizing mask translation to use Auto Layout constraints.

!!step

> lines-BannerViewController.swift=43-57

Reviewer 2: Looks like this is highlighting the wrong lines for the UIKit part
<img width="781" alt="Screenshot 2025-06-04 at 4 27 37 PM" src="https://github.com/user-attachments/assets/7eba5f75-9d78-427e-9c9f-fca59eacbcce" />

#### 5. Pin views with Auto Layout constraints and set up height constraint

Anchor your main content to the top. Place the banner view directly below, pinning its leading, trailing, and bottom to the safe area.
Set up a height constraint for the banner view (starting at 0), which will be updated when banner content loads.

{% endscrolly %}
{% endtab %}
{% tab SwiftUI %}
{% scrolly %}

```swift file=AppDelegate.swift

> import UIKit

Reviewer 2: Remove since this is the SwiftUI tab
```suggestion
```

import BrazeKit
import BrazeUI

class AppDelegate: UIResponder, UIApplicationDelegate {

>     static var braze: Braze!
> 
>     func application(
>       _ application: UIApplication,
>       didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
>     ) -> Bool {
> 
>         // Braze configuration with your SDK API key and endpoint
>         let configuration = Braze.Configuration(apiKey: "YOUR-API-TOKEN", endpoint: "YOUR-ENDPOINT")
>         configuration.logger.level = .debug
> 
>         // Initialize Braze SDK instance
>         AppDelegate.braze = Braze(configuration: configuration)
> 
>         // Request a banners refresh
>         AppDelegate.braze.banners.requestBannersRefresh(placementIds: ["top-1"])
> 
>         return true

Reviewer 2: nit: I noticed we treat `braze` as nullable in the other files like `BrazeSwiftUIView.swift`, so this would make it actually be nullable here
```suggestion
    static var braze: Braze? = nil

    func application(
      _ application: UIApplication,
      didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
    ) -> Bool {
        // Braze configuration with your SDK API key and endpoint
        let configuration = Braze.Configuration(apiKey: "YOUR-API-TOKEN", endpoint: "YOUR-ENDPOINT")
        configuration.logger.level = .debug

        // Initialize Braze SDK instance
        AppDelegate.braze = Braze(configuration: configuration)

        // Request a banners refresh
        AppDelegate.braze?.banners.requestBannersRefresh(placementIds: ["top-1"])

        return true
```

  @State var hasBannerForPlacement: Bool = false
  @State var contentHeight: CGFloat = 0

  var body: some View {
    VStack {
      Text("Your Content Here")
        .frame(maxWidth: .infinity, maxHeight: .infinity)
      if let braze = AppDelegate.braze,
        hasBannerForPlacement
      {
        BrazeBannerUI.BannerView(
          placementId: BannerSwiftUIView.bannerPlacementID,
          braze: braze,
          processContentUpdates: { result in
            switch result {
            case .success(let updates):
              if let height = updates.height {
                self.contentHeight = height
              }
            case .failure:
              return
            }
          }
        )
        .frame(height: min(contentHeight, 80))
      }
    }.onAppear {
      AppDelegate.braze?.banners.getBanner(
        for: BannerSwiftUIView.bannerPlacementID,
        { banner in
          hasBannerForPlacement = banner != nil
        }
      )
    }
  }
}

```

!!step
lines-AppDelegate.swift=15

#### 1. Enable debugging (optional)

Enable debugging while developing to make troubleshooting easier!

!!step
lines-AppDelegate.swift=21

> #### 2. Trigger a banners refresh (optional)

Reviewer 2: ```suggestion
#### 2. Trigger a banners refresh
```

---
nav_title: Integrating Banners
guide_top_header: Integrating Banners
article_title: Integrating Banners
page_order: 1
layout: scrolly
description: "A tutorial on how to integrate and refresh banner placements"
---

> # Conditionally Display In-App Messages

Reviewer 3: Maybe something like:

```suggestion
# Displaying a Banner by its Placement ID
```

{% tabs %}
{% tab Web %}
{% scrolly %}

```js file=index.js
import * as braze from "@braze/web-sdk";

braze.initialize("YOUR-API-KEY", {
  baseUrl: "YOUR-ENDPOINT",
  enableLogging: true,
});

braze.subscribeToBannersUpdates((banners) => {

>   // get this placement's banner. If it's `null` the user did not qualify for one.

Reviewer 4: ```suggestion
  // Get this placement's banner. If it's `null`, the user did not qualify for any banners.
```

Call this whenever you want to refresh the banners for any placement(s).

!!step
lines-main.html=3

#### 5. Add a container for your banner

Add a div in your HTML (e.g. `<div id="global-banner-container"></div>`) where Braze will insert the banner content.

{% endscrolly %}
{% endtab %}
{% tab iOS %}
{% subtabs %}
{% subtab UIKit %}
{% scrolly %}

```swift file=AppDelegate.swift
import SwiftUI
import BrazeKit
import BrazeUI

class AppDelegate: UIResponder, UIApplicationDelegate {
    static var braze: Braze!

    func application(
      _ application: UIApplication,
      didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
    ) -> Bool {

        // Braze configuration with your SDK API key and endpoint
        let configuration = Braze.Configuration(apiKey: "YOUR-API-TOKEN", endpoint: "YOUR-ENDPOINT")
        configuration.logger.level = .debug

        // Initialize Braze SDK instance
        AppDelegate.braze = Braze(configuration: configuration)

        // Request a banners refresh
        AppDelegate.braze.banners.requestBannersRefresh(placementIds: ["top-1"])

        return true
    }
}
```

> ```swift file=ContentView.swift

Reviewer 3: This wouldn't really be part of a UIKit example. The code shown below fits more in a SwiftUI setup. I guess it would be valid if someone was embedding a UIKit view controller within a primarily SwiftUI app, but overall, I don't think this file would be necessary for this example. We should be able to just have the view controller.

Author: Ok, so is simply removing that file enough for this UIKIT implementation to make more sense?

Reviewer 3: Yep I believe so - if I'm not mistaken, I'm not seeing it being referenced in any of the tutorial steps, so we should be able to do without it.

  lazy var contentView: UILabel = {
    let contentView = UILabel()
    contentView.text = "Your Content Here"
    contentView.textAlignment = .center
    contentView.translatesAutoresizingMaskIntoConstraints = false
    return contentView
  }()

  lazy var bannerView: BrazeBannerUI.BannerUIView = {
    var bannerView = BrazeBannerUI.BannerUIView(
      placementId: BannerViewController.bannerPlacementID,
      braze: AppDelegate.braze!,
      processContentUpdates: { [weak self] result in
        // Update layout properties when banner content has finished loading.
        DispatchQueue.main.async {
          guard let self else { return }
          switch result {
          case .success(let updates):
            if let height = updates.height {
              self.bannerView.isHidden = false
              self.bannerHeightConstraint?.constant = min(height, 80)
            }
          case .failure(let error):
            return
          }
        }
      }
    )
    bannerView.translatesAutoresizingMaskIntoConstraints = false
    bannerView.isHidden = true
    return bannerView
  }()

  override func loadView() {
    super.loadView()

>     view.backgroundColor = .white

Reviewer 3: Do we need this (or rather the `loadView` function as a whole)? Seems like an extra step that wouldn't be necessary for the scope of this example.

Author: Probably not, I was copying the swift examples we have in the repo. What does `super.loadView()` do?

  }

  override func viewDidLoad() {
    super.viewDidLoad()
    self.view.addSubview(contentView)
    self.view.addSubview(bannerView)
    bannerHeightConstraint = bannerView.heightAnchor.constraint(equalToConstant: 0)
    NSLayoutConstraint.activate([
      contentView.topAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.topAnchor),
      contentView.leadingAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.leadingAnchor),
      contentView.trailingAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.trailingAnchor),
      bannerView.topAnchor.constraint(equalTo: self.contentView.bottomAnchor),
      bannerView.leadingAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.leadingAnchor),
      bannerView.trailingAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.trailingAnchor),
      bannerView.bottomAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.bottomAnchor),
      bannerHeightConstraint!,
    ])
  }
}
```

!!step
lines-AppDelegate.swift=15

#### 1. Enable debugging (optional)

Enable debugging while developing to make troubleshooting easier!

!!step
lines-AppDelegate.swift=21

> #### 2. Trigger a banners refresh (optional)

Reviewer 3: Should this not be optional? They'll need the banner from the server in order to display it unless we want to assume that they already have it in the cache.
```suggestion
#### 2. Trigger a banners refresh
```

Author: Hmmmm for the sake of this example, yeah that's a good call

Call `requestBannersRefresh` for your placement ID after initializing Braze.
This proactively fetches the latest banner data, ensuring your app can display new content quickly after launch.

!!step
lines-BannerViewController.swift=18-32

#### 3. Initialize the Braze banner view with a processContentUpdates callback

Create your `BrazeBannerUI.BannerUIView` with the correct placement ID and Braze instance.

> The `processContentUpdates` callback will fire whenever the banner content (and height) changes.

Reviewer 3: To prevent this from being misleading, should we just say `banner height`? This wouldn't trigger if any of the banner content changes (i.e. text, colors, etc.), only just the height. We kept it generic with `updates` to accommodate possible additions in the future, like if we wanted to send the width, but for now, this should just be the height.

Author: ```suggestion
The `processContentUpdates` callback will fire whenever the banner height changes.
```

> Inside the callback, unhide the banner and update its height constraint using the provided content height (capped at 80pt for consistency).

Reviewer 3: ```suggestion
Inside the callback, unhide the banner and update its height constraint using the provided content height (capped at 80pt for this example).
```

!!step
lines-BannerViewController.swift=34-36

#### 4. Hide banner by default, enable Auto Layout

Hide the banner view until content loads and disable autoresizing mask translation to use Auto Layout constraints.

!!step
lines-BannerViewController.swift=39-42

#### 5. Configure the view hierarchy and background

Set up your controller’s background color and add both the main content label and banner view to the hierarchy.

!!step

> lines-BannerViewController.swift=50-63

Reviewer 3: To include the line with 0 default height.
```suggestion
lines-BannerViewController.swift=48-63
```

This prevents blank/white spaces in your UI when no banner is available.

!!step
lines-BannerSwiftUIView.swift=23-32

#### 6. Dynamically update banner height

Use the `processContentUpdates` callback to track banner content height as soon as it loads.
Update your SwiftUI state (`contentHeight`) and use `.frame(height:)` to control the banner’s visible height in your layout.

!!step
lines-BannerSwiftUIView.swift=34

#### 7. Limit banner height for a consistent look

Apply a `.frame(height: min(contentHeight, 80))` modifier to ensure your banner never exceeds a maximum height (e.g., 80 points), keeping the UI visually balanced regardless of content.

{% endscrolly %}
{% endsubtab %}
{% endsubtabs %}
{% endtab %}
{% tab Android %}
{% scrolly %}

```kotlin file=MainApplication.kt
import android.app.Application
import android.util.Log
import com.braze.Braze
import com.braze.configuration.BrazeConfig
import com.braze.support.BrazeLogger

public class MainApplication : Application() {
    override fun onCreate() {
        super.onCreate()

        // Turn on verbose Braze logging
        BrazeLogger.logLevel = Log.VERBOSE

>         // Configure Braze with your SDK key & endpoint

Reviewer 4: ```suggestion
        // Configure Braze with your SDK key and endpoint
```

        val config = BrazeConfig.Builder()
            .setApiKey("YOUR-API-KEY")
            .setCustomEndpoint("YOUR-ENDPOINT")
            .build()
        Braze.configure(this, config)

        // Subscribe to Banner updates
        Braze.getInstance(this)
            .subscribeToBannersUpdates { update ->
                for (banner in update.banners) {
                    Log.d("brazeBanners", "Received banner for placement: ${banner.placementId}")

>                     // any custom logic you'd like

Reviewer 4: ```suggestion
                    // Add any custom banner logic you'd like
```

                }
            }

        // Refresh placements
        Braze.getInstance(this)
            .requestBannersRefresh(
                listOf("top-1")
            )
    }
}
```

```kotlin file=MainActivity.kt
import android.os.Bundle
import androidx.activity.ComponentActivity

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

>         // Inflate the XML layout:

Reviewer 4: ```suggestion
        // Inflate the XML layout
```

        <!-- Banner placement -->
        <com.braze.ui.banners.BannerView
            android:id="@+id/banner_view_1"
            android:layout_width="match_parent"
            android:layout_height="wrap_content"
            app:placementId="top-1" />

        <!-- ...the rest of your activity layout -->

    </LinearLayout>
</ScrollView>
```

!!step
lines-MainApplication.kt=12

#### 1. Enable debugging (optional)

Enable debugging while developing to make troubleshooting easier!

!!step
lines-MainApplication.kt=21-28

#### 2. Subscribe to banner updates

Use `BrazeInAppMessageManager` to set a custom listener that will intercept messages before they're displayed.

!!step
lines-MainApplication.kt=30-34

#### 3. Refresh placements

Invoke `requestBannersRefresh()` to fetch the latest banner content from Braze for the specified placement ID(s).

!!step
lines-MainActivity.kt=7-8

#### 4. Inflate the XML layout containing your BannerView

In MainActivity, call `setContentView(R.layout.banners)` so that the activity loads the layout file where your BannerView will be defined. Optionally, you can also use `Banner(placementId = "placement-id")` if using Jetpack Compose.

!!step

> lines-banners.xml=15-19

Author: Ok, will remove this step! Thanks for clarifying
